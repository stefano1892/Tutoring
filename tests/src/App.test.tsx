import { render } from '@testing-library/react';
import { getDegrees } from './api';
import userEvent from '@testing-library/user-event';
import App from './App';

vi.mock('./api')

describe('App', () => {
    const getDegreesMock = vi.mocked(getDegrees)
    
    /*
        test('at the beginning, counter value is 0', () => {
            const { getByText } = render(<App />)
    
            expect(getByText('Contatore: 0')).toBeDefined()
        })
    
        test('after clicking on button, counter value is 1', async () => {
            const { getByText, getByRole } = render(<App />)
    
            const incrementButton = getByRole('button', { name: 'Incrementa' })
            await userEvent.click(incrementButton)
    
            expect(getByText('Contatore: 1')).toBeDefined()
        })
    */

    test('after clicking the button n times, counter value is n', async () => {
        const clickTimes = 5
        const { getByText, getByRole } = render(<App />)

        const incrementButton = getByRole('button', { name: 'Incrementa' })
        for (let i = 0; i < clickTimes; i++) {
            await userEvent.click(incrementButton)
        }

        expect(getByText(`Contatore: ${clickTimes}`)).toBeDefined()
    })

    test('match snapshot - initial state', () => {
        const { asFragment } = render(<App />)
        expect(asFragment()).toMatchSnapshot()
    })

    test('match snapshot - clicking on button 1 time', async () => {
        const { asFragment, getByRole } = render(<App />)

        const incrementButton = getByRole('button', { name: 'Incrementa' })
        await userEvent.click(incrementButton)

        expect(asFragment()).toMatchSnapshot()
    })

    describe('get degree api works correctly', () => {
        beforeEach(() => {
            getDegreesMock.mockResolvedValue(21)
        })

        test('match snapshot - expect degree to be invokes correctly', async () => {
            const { asFragment, getByRole } = render(<App />)
            const temperatureButton = getByRole('button', {name: 'Trova temperatura'})

            await userEvent.click(temperatureButton)
            expect(getDegreesMock).toHaveBeenCalledWith('Milano')
            expect(asFragment()).toMatchSnapshot()
        })

        test('match snapshot - expect degree to be the same if button clicked twice', async () => {
            const { asFragment, getByRole } = render(<App />)
            const temperatureButton = getByRole('button', {name: 'Trova temperatura'})

            await userEvent.click(temperatureButton)
            await userEvent.click(temperatureButton)
            expect(getDegreesMock).toHaveBeenCalledWith('Milano')
            expect(asFragment()).toMatchSnapshot()
        })
    })

    test('match snapshot - expect degree to be -1 when error happens', async () => {
        getDegreesMock.mockRejectedValue(new Error('FOOOOOO'))

        const { asFragment, getByRole } = render(<App />)
        const temperatureButton = getByRole('button', {name: 'Trova temperatura'})

        await userEvent.click(temperatureButton)
        expect(getDegreesMock).toHaveBeenCalledWith('Milano')
        expect(asFragment()).toMatchSnapshot()
    })
})