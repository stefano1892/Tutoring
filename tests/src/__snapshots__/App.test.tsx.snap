// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > get degree api works correctly > match snapshot - expect degree to be invokes correctly 1`] = `
<DocumentFragment>
  <button>
    Incrementa
  </button>
  <p>
    Contatore: 0
  </p>
  <button>
    Trova temperatura
  </button>
  <p>
    Temperatura: 21
  </p>
</DocumentFragment>
`;

exports[`App > get degree api works correctly > match snapshot - expect degree to be the same if button clicked twice 1`] = `
<DocumentFragment>
  <button>
    Incrementa
  </button>
  <p>
    Contatore: 0
  </p>
  <button>
    Trova temperatura
  </button>
  <p>
    Temperatura: 21
  </p>
</DocumentFragment>
`;

exports[`App > match snapshot - clicking on button 1 time 1`] = `
<DocumentFragment>
  <button>
    Incrementa
  </button>
  <p>
    Contatore: 1
  </p>
  <button>
    Trova temperatura
  </button>
  <p>
    Temperatura: 0
  </p>
</DocumentFragment>
`;

exports[`App > match snapshot - expect degree to be -1 when error happens 1`] = `
<DocumentFragment>
  <button>
    Incrementa
  </button>
  <p>
    Contatore: 0
  </p>
  <button>
    Trova temperatura
  </button>
  <p>
    Temperatura: -1
  </p>
</DocumentFragment>
`;

exports[`App > match snapshot - initial state 1`] = `
<DocumentFragment>
  <button>
    Incrementa
  </button>
  <p>
    Contatore: 0
  </p>
  <button>
    Trova temperatura
  </button>
  <p>
    Temperatura: 0
  </p>
</DocumentFragment>
`;
